File: advent-2023/python/q23.py
from collections import deque
import heapq

def find_pt(row, grid):
    for col in range(len(grid[row])):
        if grid[row][col] == '.':
            return (row,col)

def step_slope(tile, row, col):
    match tile:
        case '^':
            return (row-1, col)
        case '>':
            return (row, col+1)
        case 'v':
            return (row+1, col)
        case '<':
            return (row, col-1)
    print("ERR")
    exit(1)
    
def print_grid(visited, grid):
    for row in range(len(grid)):
        s = ""
        for col in range(len(grid[0])):
            if (row,col) in visited:
                s += "@"
            else:
                s += grid[row][col]
        print(s)

def part1(grid):
    start = find_pt(0, grid)
    end = find_pt(len(grid) - 1, grid)
    visited = {}
    heap = ([(0, start)])
    max_distance = 0

    while heap:
        distance, key = heapq.heappop(heap)

        if key == end:
            print(distance)
            max_distance  = min(max_distance, distance)
            continue
        
        if key in visited and distance < visited[key]:
            continue
        visited[key] = distance
        
        row,col = key
        # print(key)
        for dir in [(0,1),(0,-1),(1,0),(-1,0)]:
            nrow, ncol= row + dir[0], col + dir[1]
            if not(0 <= nrow < len(grid) and 0 <= ncol < len(grid[0])):
                continue 
            tile = grid[nrow][ncol]
            # print(nrow, ncol, tile)
            if tile == '#':
                continue
            heapq.heappush(heap, (distance - 1, (nrow,ncol)))
            # if tile == '.':
            # elif tile in "^>v<":
            #     skey = step_slope(tile,nrow,ncol)
            #     if skey in visited:
            #         continue
            #     new_visited.add(skey)
            #     stack.append((skey, new_visited, distance + 2))
    print("max_distance", max_distance)

def solution():
    filename = "./inputs/q23.txt"
    grid = open(filename).read().split('\n')
    part1(grid)
    pass

solution()

# def part1(grid):
#     start = find_pt(0, grid)
#     end = find_pt(len(grid) - 1, grid)
#     print(start, end)
#     visited = set()
#     visited.add(start)
#     stack = [(start, visited, 0 )]
#     max_distance = 0
#     while stack:
#         key, visited, distance = stack.pop()

#         if key == end:
#             print(distance)
#             max_distance  = max(max_distance, distance)
#             continue
        
#         row,col = key
#         # print(key)
#         for dir in [(0,1),(0,-1),(1,0),(-1,0)]:
#             nrow, ncol= row + dir[0], col + dir[1]
#             if not(0 <= nrow < len(grid) and 0 <= ncol < len(grid[0])):
#                 continue 
#             tile = grid[nrow][ncol]
#             # print(nrow, ncol, tile)
#             if tile == '#' or (nrow,ncol) in visited:
#                 continue
#             new_visited = {v for v in visited} 
#             new_visited.add((nrow,ncol))
#             stack.append(((nrow,ncol), new_visited, distance + 1))
#             # if tile == '.':
#             # elif tile in "^>v<":
#             #     skey = step_slope(tile,nrow,ncol)
#             #     if skey in visited:
#             #         continue
#             #     new_visited.add(skey)
#             #     stack.append((skey, new_visited, distance + 2))
#     print("max_distance", max_distance)

File: advent-2023/c/program/q23.c
#include "call_lib.h"

// enum { GRID_SIZE = 141 };
enum { GRID_SIZE = 23 };
static const int DIRECTIONS[4][2] = {{-1,0},{0,1},{1,0},{0,-1}};
static int MAX_STEPS = 0;

char** create_grid(char* input) {
    char** grid = malloc(GRID_SIZE * sizeof(char*));
    char* rows = malloc(GRID_SIZE * GRID_SIZE * sizeof(char));
    size_t i, j;
    
    for (i=0; i < GRID_SIZE; i++) {
        grid[i] = rows + i * GRID_SIZE;
    }

    for (i=0; i < GRID_SIZE; i++) {
        for (j=0; j < GRID_SIZE; j++) {
            grid[i][j] = *input++;
        }
        input++;
    }

    return grid;
}

int find_direction(char tile) {
    size_t i;
    char pdir[] = "^>v<";
    for(i=0; i < 4; i++) {
        if (tile == pdir[i]) return i;
    }
    return -1;
}

void traverse_grid(int row, int col, char** grid, bool visited[][GRID_SIZE], int steps) {
    char tile; size_t i; int pdir;
    if (row < 0 || row >= GRID_SIZE || col < 0 || col >= GRID_SIZE) return;
    
    tile = grid[row][col];

    if (tile == '#' || visited[row][col]) return;

    if (row == GRID_SIZE -1 && col == GRID_SIZE - 2) {
        printf("steps: %d\n", steps);
        MAX_STEPS = MAX(MAX_STEPS, steps);
        return;
    }

    // part 1
    // pdir = find_direction(tile);
    // if (pdir != -1) {
    //     int nrow = row + DIRECTIONS[pdir][0];
    //     int ncol = col + DIRECTIONS[pdir][1];
    //     visited[row][col] = TRUE;
    //     traverse_grid(nrow, ncol, grid, visited, steps + 1);
    //     visited[row][col] = FALSE;
    //     return;
    // } 

    visited[row][col] = TRUE;
    for (i=0; i< 4; i++) {
        int nrow = row + DIRECTIONS[i][0];
        int ncol = col + DIRECTIONS[i][1];
        traverse_grid(nrow, ncol, grid, visited, steps + 1);
    }
    visited[row][col] = FALSE;
}

void part1(char *input) {
    char** grid = create_grid(input);
    bool visited[GRID_SIZE][GRID_SIZE] = {0};
    traverse_grid(0, 1, grid, visited, 0);
    printf("max: %d\n", MAX_STEPS);
}

void part2(char *input) {}

SOLUTION("./inputs/q23.txt")
